src/main/java/io/jenkins/plugins/checks/github/GitHubSCMSourceChecksContext.java

```diff
@@ -45,7 +48,8 @@
      */
     GitHubSCMSourceChecksContext(final Job<?, ?> job, final String jobURL, final SCMFacade scmFacade) {
         super(job, jobURL, scmFacade);
-        sha = resolveHeadSha(job);
+        this.run = null;
```

Comment:
This PMD warning is annoying 😤

Since we don't have a null object pattern for the run, I can't find a way to solve this other than shooting the message.

Reply:
I could make run a `Optional<Run<?, ?>>`, but I gather that is frowned upon too?

Reply:
Yes, right! I always thought possess an optional inside a class makes another warnings, can't remember where I saw this...

Reply:
Well IntelliJ doesn’t like it, and IntelliJ is right about _everything_ 😉

Reply:
Well, I think the warning makes sense here. We introduced two constructors to avoid the problem of assigning ˋnullˋ to the run. If you keep it this way you can remove the constructor as well

Reply:
I'm not sure I follow - which constructor can we remove?

Reply:
In the beginning of the project we had one constructor for this class that sets all not used values to `null`, so if it makes sense we can combine the two constructors.

Reply:
Ah, so something like https://gist.github.com/mrginglymus/34bf66256e5af009be7e7c568cba9401 ?

That moves (and duplicates) the logic for creating the context from a run into the calling code, which I don't think looks as nice.

Reply:
I guess we could have static `fromRun` and `fromJob` methods...

Reply:
@uhafner would you prefer something like:

```java
    static GitHubSCMSourceChecksContext fromRun(final Run<?, ?> run, final String runURL, final SCMFacade scmFacade) {
        return new GitHubSCMSourceChecksContext(run.getParent(), run, runURL, scmFacade);
    }

    static GitHubSCMSourceChecksContext fromJob(final Job<?, ?> job, final String runURL, final SCMFacade scmFacade) {
        return new GitHubSCMSourceChecksContext(job, null, runURL, scmFacade);
    }

    /**
     * Creates a {@link GitHubSCMSourceChecksContext} according to the run. All attributes are computed during this period.
     *
     * @param job
     *         a GitHub Branch Source project
     * @param run
     *         a run of a GitHub Branch Source project
     * @param runURL
     *         the URL to the Jenkins run
     * @param scmFacade
     *         a facade for Jenkins SCM
     */
    private GitHubSCMSourceChecksContext(final Job<?, ?> job, @CheckForNull final Run<?, ?> run, final String runURL, final SCMFacade scmFacade) {
        super(job, runURL, scmFacade);
        this.run = run;
        this.sha = Optional.ofNullable(run).map(this::resolveHeadSha).orElse(resolveHeadSha(job));
    }
```

Reply:
An alternative way, a bigger refactor: we move all the action related methods into the publisher directly, and let the publisher possess the run. After all, I don't feel comfortable to see the context holds a job and a run (which is only used for this action) at the same time.

Reply:
I think we don't have to have the static methods, after all, the context is only used inside this package; they just add up the complexity.

Reply:
The advantage behind having this way (preserved from having two constructors) is that it greatly reduces the chances of creating a `GitHubSCMSourceChecksContext` where the run does not belong to the job. I appreciate, as you say, that it's only used within the package but it still feels like a smell to have to duplicate the `run.getParent(), run` logic throughout the test code.